presto-main/src/main/java/com/facebook/presto/sql/MaterializedViewUtils.java

```diff
@@ -0,0 +1,96 @@
+     * all the N base table columns that one materialized view column maps to.
+     * It calls on {@link MaterializedViewColumnMappingExtractor} to get all base table columns mapped by join, union, etc,
+     * and then uses them to expand the 1-to-1 column mapping derived from {@code analysis} to the 1-to-N column mapping.
+     * <p>
+     * For example, given SELECT column_a AS column_x FROM table_a JOIN table_b ON (table_a.column_a = table_b.column_b),
+     * the 1-to-1 column mapping from {@code analysis} is column_x -> table_a.column_a. Mapped base table columns are
+     * [table_a.column_a, table_b.column_b]. Then it will return a 1-to-N column mapping column_x -> {table_a.column_a, table_b.column_b}.
+     */
+    public static Map<String, Map<SchemaTableName, String>> computeMaterializedViewToBaseTableColumnMappings(Analysis analysis)
+    {
+        checkState(
+                analysis.getStatement() instanceof CreateMaterializedView,
+                "Only support the computation of column mappings when analyzing CreateMaterializedView");
+
+        ImmutableMap.Builder<String, Map<SchemaTableName, String>> fullColumnMapping = ImmutableMap.builder();
+
+        Query viewQuery = ((CreateMaterializedView) analysis.getStatement()).getQuery();
+        Map<String, Map<SchemaTableName, String>> originalColumnMapping = getOriginalColumnMappingFromAnalysis(viewQuery, analysis);
+
+        List<List<TableColumn>> mappedBaseColumns = MaterializedViewColumnMappingExtractor.extractMappedBaseColumns(analysis);
+
+        for (Map.Entry<String, Map<SchemaTableName, String>> entry : originalColumnMapping.entrySet()) {
+            String viewColumn = entry.getKey();
+            Map<SchemaTableName, String> originalBaseColumns = entry.getValue();
+
+            Map<SchemaTableName, String> fullBaseColumns = new HashMap<>(originalBaseColumns);
+
+            originalBaseColumns.forEach((originalBaseTable, originalBaseColumn) -> {
+                TableColumn originalTableColumn = new TableColumn(originalBaseTable, originalBaseColumn);
+                mappedBaseColumns.forEach(linkedTableColumnPair -> {
+                    if (originalTableColumn.equals(linkedTableColumnPair.get(0))) {
+                        fullBaseColumns.put(linkedTableColumnPair.get(1).getTableName(), linkedTableColumnPair.get(1).getColumnName());
+                    }
+                    else if (originalTableColumn.equals(linkedTableColumnPair.get(1))) {
+                        fullBaseColumns.put(linkedTableColumnPair.get(0).getTableName(), linkedTableColumnPair.get(0).getColumnName());
+                    }
+                });
+            });
+
+            fullColumnMapping.put(viewColumn, ImmutableMap.copyOf(fullBaseColumns));
+        }
+
+        return fullColumnMapping.build();
+    }
+
+    private static Map<String, Map<SchemaTableName, String>> getOriginalColumnMappingFromAnalysis(Node viewQuery, Analysis analysis)
```

Comment:
It is a little bit hard to figure out what this function does from its name. May be better to rename it to
getOutputFieldToOriginalColumnMap

Reply:
Revised